daffodil-codegen-c/src/main/scala/org/apache/daffodil/codegen/c/generators/BinaryValueCodeGenerator.scala

```diff
@@ -93,11 +141,94 @@ trait BinaryValueCodeGenerator {
 
     val initERDStatement = ""
     val parseStatement =
-      s"""$indent1$indent2    parse_validate_fixed($field == $fixed, "$localName", pstate);
-         |$indent1$indent2    if (pstate->error) return;""".stripMargin
+      s"""$indent1$indent2    validate_fixed_attribute($field == $fixed, "$localName", &pstate->pu);
+         |$indent1$indent2    if (pstate->pu.error) return;""".stripMargin
+    val unparseStatement =
+      s"""$indent1$indent2    validate_fixed_attribute($field == $fixed, "$localName", &ustate->pu);
+         |$indent1$indent2    if (ustate->pu.error) return;""".stripMargin
+    cgState.addSimpleTypeStatements(initERDStatement, parseStatement, unparseStatement)
+  }
+
+  // Generate C code to validate a primitive element matches one of its enumeration values.
+  // May be replaced by a more specialized function in another trait for some types.
+  private def valueValidateEnumeration(
+    e: ElementBase,
+    deref: String,
+    enums: Array[String],
+    cgState: CodeGeneratorState,
+  ): Unit = {
+    val indent1 = if (cgState.hasChoice) INDENT else NO_INDENT
+    val indent2 = if (deref.nonEmpty) INDENT else NO_INDENT
+    val localName = cgState.cName(e)
+    val field = s"instance->$localName$deref"
+
+    // Extra initialization only for hexBinary enumerations
+    val arraysName = s"arrays_$localName"
+    val hexEnums = enums.zipWithIndex.map { case (s, index) =>
+      s"{$arraysName[$index], ${s.length / 2}, false}"
+    }
+    val hexEnumsInit = enums.map(_.grouped(2).map("0x" + _).mkString("{", ", ", "}"))
+    val enumsLenMax = enums.map(_.length / 2).max
+    val arraysInit =
+      s"""$indent1$indent2    uint8_t $arraysName[][$enumsLenMax] = ${hexEnumsInit.mkString(
+          "{",
+          ", ",
+          "}",
+        )};\n"""
+
+    val (enumsInit, extraInit, primType, valType, fieldArg) = e.optPrimType.get match {
+      case PrimType.Double | PrimType.Float => (enums, "", "double", "floatpt", field)
+      case PrimType.HexBinary => (hexEnums, arraysInit, "HexBinary", "hexbinary", s"&$field")
+      case _ => (enums, "", "int64_t", "integer", field)
```

Comment:
I don't think I ever checked dates, times, or arbitrary length integers longer than 64 bits. I am sure they would fail somewhere else too since the code generator doesn't support them at all (it won't be able to find the right primitive C type, for example).